### ISSUE TITLE
Can't build in no_std environment

### ISSUE THREAD

**Issue by ethindp** [paritytech/wasmi#204]
The latest crates.io build no longer works with the core feature. num-traits is required by one of the dependent crates or by wasmi itself, and the std feature is not turned off, causing E0463 ("can't find crate for `std`"). This is how I'm including it:
```
wasmi = {version = "*", default-features=false, features=["core"]}
```

**Comment by pepyakin** [comment]
That's interesting, did you use nightly? What is the error?

**Comment by thiolliere** [comment]
This seems to be solved on master but not on version in crate.io

```
[thiolliere@localhost ressai]$ cargo +nightly check --no-default-features 
   Compiling byteorder v1.3.2
    Checking parity-wasm v0.31.3
    Checking hashbrown v0.1.8
    Checking wasmi-validation v0.1.0
    Checking wasmi v0.5.0
error[E0433]: failed to resolve: maybe a missing crate `std`?
 --> /home/thiolliere/.cargo/registry/src/github.com-1ecc6299db9ec823/wasmi-0.5.0/src/memory/mmap_bytebuf.rs:8:5
  |
8 | use std::ptr::{self, NonNull};
  |     ^^^ maybe a missing crate `std`?

error[E0432]: unresolved import `std`
 --> /home/thiolliere/.cargo/registry/src/github.com-1ecc6299db9ec823/wasmi-0.5.0/src/memory/mmap_bytebuf.rs:8:5
  |
8 | use std::ptr::{self, NonNull};
  |     ^^^ maybe a missing crate `std`?

error[E0432]: unresolved import `std`
 --> /home/thiolliere/.cargo/registry/src/github.com-1ecc6299db9ec823/wasmi-0.5.0/src/memory/mmap_bytebuf.rs:9:5
  |
9 | use std::slice;
  |     ^^^ maybe a missing crate `std`?

error[E0433]: failed to resolve: use of undeclared type or module `NonNull`
  --> /home/thiolliere/.cargo/registry/src/github.com-1ecc6299db9ec823/wasmi-0.5.0/src/memory/mmap_bytebuf.rs:66:27
   |
66 |                 let ptr = NonNull::new(ptr_or_err as *mut u8).ok_or("mmap returned 0")?;
   |                           ^^^^^^^ use of undeclared type or module `NonNull`

error[E0412]: cannot find type `NonNull` in this scope
  --> /home/thiolliere/.cargo/registry/src/github.com-1ecc6299db9ec823/wasmi-0.5.0/src/memory/mmap_bytebuf.rs:15:10
   |
15 |     ptr: NonNull<u8>,
   |          ^^^^^^^ not found in this scope
help: possible candidate is found in another module, you can import it into scope
   |
8  | use core::ptr::NonNull;
   |

error: aborting due to 5 previous errors

Some errors have detailed explanations: E0412, E0432, E0433.
For more information about an error, try `rustc --explain E0412`.
error: Could not compile `wasmi`.

```

**Comment by ethindp** [comment]
Yes, I used nightly. I would use the git hub repository but I don't
want to risk random instability.

On 8/27/19, thiolliere <notifications@github.com> wrote:
> This seems to be solved on master but not on version in crate.io
>
> ```
> [thiolliere@localhost ressai]$ cargo +nightly check --no-default-features
>    Compiling byteorder v1.3.2
>     Checking parity-wasm v0.31.3
>     Checking hashbrown v0.1.8
>     Checking wasmi-validation v0.1.0
>     Checking wasmi v0.5.0
> error[E0433]: failed to resolve: maybe a missing crate `std`?
>  -->
> /home/thiolliere/.cargo/registry/src/github.com-1ecc6299db9ec823/wasmi-0.5.0/src/memory/mmap_bytebuf.rs:8:5
>   |
> 8 | use std::ptr::{self, NonNull};
>   |     ^^^ maybe a missing crate `std`?
>
> error[E0432]: unresolved import `std`
>  -->
> /home/thiolliere/.cargo/registry/src/github.com-1ecc6299db9ec823/wasmi-0.5.0/src/memory/mmap_bytebuf.rs:8:5
>   |
> 8 | use std::ptr::{self, NonNull};
>   |     ^^^ maybe a missing crate `std`?
>
> error[E0432]: unresolved import `std`
>  -->
> /home/thiolliere/.cargo/registry/src/github.com-1ecc6299db9ec823/wasmi-0.5.0/src/memory/mmap_bytebuf.rs:9:5
>   |
> 9 | use std::slice;
>   |     ^^^ maybe a missing crate `std`?
>
> error[E0433]: failed to resolve: use of undeclared type or module `NonNull`
>   -->
> /home/thiolliere/.cargo/registry/src/github.com-1ecc6299db9ec823/wasmi-0.5.0/src/memory/mmap_bytebuf.rs:66:27
>    |
> 66 |                 let ptr = NonNull::new(ptr_or_err as *mut
> u8).ok_or("mmap returned 0")?;
>    |                           ^^^^^^^ use of undeclared type or module
> `NonNull`
>
> error[E0412]: cannot find type `NonNull` in this scope
>   -->
> /home/thiolliere/.cargo/registry/src/github.com-1ecc6299db9ec823/wasmi-0.5.0/src/memory/mmap_bytebuf.rs:15:10
>    |
> 15 |     ptr: NonNull<u8>,
>    |          ^^^^^^^ not found in this scope
> help: possible candidate is found in another module, you can import it into
> scope
>    |
> 8  | use core::ptr::NonNull;
>    |
>
> error: aborting due to 5 previous errors
>
> Some errors have detailed explanations: E0412, E0432, E0433.
> For more information about an error, try `rustc --explain E0412`.
> error: Could not compile `wasmi`.
>
> ```
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/paritytech/wasmi/issues/204#issuecomment-525264415


-- 
Signed,
Ethin D. Probst

**Comment by pepyakin** [comment]
Ah, this can be solved by using the `vec_memory` feature. However, I am not sure why, I believe that it should be enabled automatically if compiled for `no_std`

**Comment by tomaka** [comment]
This issue is still relevant and is blocked on https://github.com/rust-num/num-bigint/issues/96.

**Comment by pepyakin** [comment]
I think this should be fixed with #230.
